datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/counter.py

```diff
@@ -421,27 +424,31 @@ def __init__(
                 )
                 raise type(e)(error) from None
 
-    def handle_counter_value_error(self, error, instance=None):
+    def handle_counter_value_error(self, error):
         # Counter requires at least 2 data points to return a meaningful value, see:
         # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhgetformattedcountervalue#remarks
         #
         # https://github.com/mhammond/pywin32/blob/main/win32/src/PyWinTypesmodule.cpp#L278
-        if error.winerror != PDH_NO_DATA:
```

Comment:
I wonder if we should keep the `PDH_NO_DATA` check. With a wildcard on the instance `PdhGetFormattedCounterArray` and `PdhGetFormatterCounterValue` can return `PDH_NO_DATA` if `CollectQueryData` has not been called yet. If they has been called once it returns `PDH_CSTATUS_INVALID_DATA`.
```python
>>> q = win32pdh.OpenQuery()
>>> h = win32pdh.AddEnglishCounter(q, "\\Processor(*)\\% Idle Time")
>>> hex(h)
'0x24cfb7baa30'
>>> win32pdh.GetFormattedCounterArray(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-2147481643, 'PdhGetFormattedCounterArray', 'No data to return.')
>>> win32pdh.GetFormattedCounterValue(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-2147481643, 'GetFormattedCounterValue', 'No data to return.')
>>> win32pdh.CollectQueryData(q)
>>> win32pdh.GetFormattedCounterArray(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-1073738822, 'PdhGetFormattedCounterArray', 'The returned data is not valid.')
>>> win32pdh.GetFormattedCounterValue(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-1073738822, 'GetFormattedCounterValue', 'The returned data is not valid.'
>>> win32pdh.CollectQueryData(q)
>>> win32pdh.GetFormattedCounterArray(h, win32pdh.PDH_FMT_DOUBLE)
{'0': 99.83076192417965, '1': 99.85407178455068, '_Total': 99.84241685436517}
>>> win32pdh.GetFormattedCounterValue(h, win32pdh.PDH_FMT_DOUBLE)
(542180608, 99.83076192417965)
```
I recognize though that as the code is now `PdhCollectQueryData` is guaranteed to be called once before `collect()`, so this case shouldn't be hit, so its up to you.

For completeness here's a snippet showing that with a specific instance both `PdhGetFormattedCounterArray` and `PdhGetFormattedCounterValue` return `PDH_INVALID_DATA`.
```python
>>> q = win32pdh.OpenQuery()
>>> h = win32pdh.AddEnglishCounter(q, "\\Processor(_Total)\\% Idle Time")
>>> win32pdh.GetFormattedCounterArray(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-1073738810, 'PdhGetFormattedCounterArray', 'The data is not valid.')
>>> win32pdh.GetFormattedCounterValue(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-1073738810, 'GetFormattedCounterValue', 'The data is not valid.')
>>> win32pdh.CollectQueryData(q)
>>> win32pdh.GetFormattedCounterArray(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-1073738810, 'PdhGetFormattedCounterArray', 'The data is not valid.')
>>> win32pdh.GetFormattedCounterValue(h, win32pdh.PDH_FMT_DOUBLE)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (-1073738810, 'GetFormattedCounterValue', 'The data is not valid.')
```

Reply:
Thank you for detailed information. This decision for this particular piece of code is "easy" though. It is actual regression. Here is [original piece of code](https://github.com/DataDog/integrations-core/blob/8e5ab3b1d3ad7bd1354bdfc4d0082b595dcda2c2/datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/counter.py#L323)
```
    def handle_counter_value_error(self, error, instance=None):
        # Counter requires at least 2 data points to return a meaningful value, see:
        # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhgetformattedcountervalue#remarks
        #
        # https://mhammond.github.io/pywin32/error.html
        if error.strerror != 'The data is not valid.':
            raise
``` 
I improved it to use a constant instead of string comparison. Unfortunately I did not use the proper error code. I thought that I had tested but I made a mistake somewhere. In addition, multi-instance counters generate a different error.

In addition I do not like even this piece of code since it may be a bit too overzealous error swallowing which I want to improve later. As I have mentioned we need to swallow errors only once, but this piece of code does not account for this.